File: .cursor/skills/common/cursor-inventory/scripts/inventory.mjs
#!/usr/bin/env node
/** Scan .cursor/ for skills, slash commands, and subagents; emit markdown tables. */
import { readFileSync, writeFileSync, mkdirSync, readdirSync, statSync } from "fs";
import { join, relative, resolve, dirname, sep } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, "..", "..", "..", "..", "..");

const FRONTMATTER_RE = /^---\s*\n([\s\S]*?)\n---/;
const FIELD_RE = /^([\w-]+):\s*(.+)$/gm;
const SKILL_LINK_RE = /\.cursor\/skills\/[^`\s]+\/SKILL\.md/gi;

function parseFrontmatter(text) {
  const match = text.match(FRONTMATTER_RE);
  if (!match) return {};
  const fields = {};
  let m;
  const block = match[1];
  FIELD_RE.lastIndex = 0;
  while ((m = FIELD_RE.exec(block)) !== null) {
    let value = m[2].trim();
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    }
    fields[m[1]] = value;
  }
  return fields;
}

function truncate(text, limit = 200) {
  const flat = text.replace(/\s+/g, " ").trim();
  return flat.length <= limit ? flat : flat.slice(0, limit - 1).trimEnd() + "…";
}

function fallbackDescription(text) {
  const body = text.replace(FRONTMATTER_RE, "").trim();
  for (const line of body.split("\n")) {
    const t = line.trim();
    if (t.startsWith("#")) return truncate(t.replace(/^#+\s*/, ""));
    if (t) return truncate(t);
  }
  return "(no description)";
}

function walk(dir, suffix, out = []) {
  for (const name of readdirSync(dir)) {
    const full = join(dir, name);
    if (statSync(full).isDirectory()) walk(full, suffix, out);
    else if (name === suffix || full.endsWith(suffix)) out.push(full);
  }
  return out;
}

function toPosix(p) {
  return p.split(sep).join("/");
}

function resolveSubagentId(filePath) {
  const stem = filePath.replace(/\.(md|mdc)$/, "").split(/[/\\]/).pop();
  return stem.startsWith("specialist-") ? stem.slice("specialist-".length) : stem;
}

function findSkillLink(text) {
  const match = text.match(SKILL_LINK_RE);
  if (!match) return "-";
  const parts = match[0].split("/");
  const idx = parts.indexOf("SKILL.md");
  return idx >= 1 ? parts[idx - 1] : "-";
}

function relCursorPath(absPath) {
  return toPosix(relative(repoRoot, absPath));
}

function collectSkills(cursorDir) {
  const rows = new Map();
  const skillDir = join(cursorDir, "skills");
  for (const file of walk(skillDir, "SKILL.md")) {
    const text = readFileSync(file, "utf8");
    const meta = parseFrontmatter(text);
    const id = meta.name || file.split(sep).slice(-2, -1)[0];
    rows.set(id, {
      id: `/${id}`,
      path: relCursorPath(file),
      description: truncate(meta.description || fallbackDescription(text)),
    });
  }
  return [...rows.values()].sort((a, b) => a.id.localeCompare(b.id));
}

function collectCommands(cursorDir) {
  const cmdDir = join(cursorDir, "commands");
  let files = [];
  try {
    files = readdirSync(cmdDir)
      .filter((f) => f.endsWith(".md"))
      .map((f) => join(cmdDir, f));
  } catch {
    return [];
  }
  return files
    .map((file) => {
      const text = readFileSync(file, "utf8");
      const meta = parseFrontmatter(text);
      const stem = file.split(sep).pop().replace(/\.md$/, "");
      return {
        id: `/${meta.name || stem}`,
        path: relCursorPath(file),
        description: truncate(meta.description || fallbackDescription(text)),
        linked: findSkillLink(text),
      };
    })
    .sort((a, b) => a.id.localeCompare(b.id));
}

function collectSubagents(cursorDir) {
  const agentsDir = join(cursorDir, "agents");
  const seen = new Set();
  const rows = new Map();
  for (const name of readdirSync(agentsDir)) {
    if (!name.endsWith(".md") && !name.endsWith(".mdc")) continue;
    const file = join(agentsDir, name);
    const key = toPosix(relative(repoRoot, file)).toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    const text = readFileSync(file, "utf8");
    const meta = parseFrontmatter(text);
    const id = meta.name || resolveSubagentId(name);
    rows.set(id, {
      id,
      path: relCursorPath(file),
      description: truncate(meta.description || fallbackDescription(text)),
      invoke: "Task `subagent_type`",
    });
  }
  return [...rows.values()].sort((a, b) => a.id.localeCompare(b.id));
}

function renderTable(headers, rows) {
  if (!rows.length) return "_None found._\n";
  const lines = [
    `| ${headers.join(" | ")} |`,
    `| ${headers.map(() => "---").join(" | ")} |`,
    ...rows.map((row) => `| ${row.map((c) => String(c).replace(/\|/g, "\\|")).join(" | ")} |`),
  ];
  return lines.join("\n") + "\n";
}

function buildMarkdown() {
  const cursorDir = join(repoRoot, ".cursor");
  const skills = collectSkills(cursorDir);
  const commands = collectCommands(cursorDir);
  const subagents = collectSubagents(cursorDir);
  const today = new Date().toISOString().slice(0, 10);

  return [
    "# Cursor skills and subagents inventory",
    "",
    `> Generated ${today} from \`.cursor/\``,
    "",
    "## Skills",
    "",
    renderTable(
      ["ID", "Path", "Description"],
      skills.map((r) => [r.id, r.path, r.description])
    ),
    "## Slash commands",
    "",
    renderTable(
      ["ID", "Path", "Description", "Linked skill"],
      commands.map((r) => [r.id, r.path, r.description, r.linked])
    ),
    "## Subagents",
    "",
    renderTable(
      ["ID", "Path", "Description", "Invoke via"],
      subagents.map((r) => [r.id, r.path, r.description, r.invoke])
    ),
    "## Summary",
    "",
    `- **Skills:** ${skills.length}`,
    `- **Slash commands:** ${commands.length}`,
    `- **Subagents:** ${subagents.length}`,
    "- **Scope:** `.cursor/skills/`, `.cursor/commands/`, `.cursor/agents/`",
    "",
  ].join("\n");
}

const outIdx = process.argv.indexOf("--out");
const outPath = outIdx >= 0 ? process.argv[outIdx + 1] : null;
const md = buildMarkdown();

if (outPath) {
  mkdirSync(resolve(outPath, ".."), { recursive: true });
  writeFileSync(outPath, md, "utf8");
  console.error(`Wrote ${outPath}`);
} else {
  process.stdout.write(md);
}

File: .cursor/skills/common/cursor-inventory/scripts/inventory.py
#!/usr/bin/env python3
"""Scan .cursor/ for skills, slash commands, and subagents; emit markdown tables."""

from __future__ import annotations

import argparse
import re
import sys
from datetime import date
from pathlib import Path


FRONTMATTER_RE = re.compile(r"^---\s*\n(.*?)\n---", re.DOTALL)
FIELD_RE = re.compile(r"^(\w[\w-]*):\s*(.+)$", re.MULTILINE)
SKILL_LINK_RE = re.compile(
    r"\.cursor/skills/[^`\s]+/SKILL\.md", re.IGNORECASE
)


def repo_root_from_script() -> Path:
    # scripts/ -> cursor-inventory/ -> common/ -> skills/ -> .cursor/ -> repo
    return Path(__file__).resolve().parents[5]


def parse_frontmatter(text: str) -> dict[str, str]:
    match = FRONTMATTER_RE.match(text)
    if not match:
        return {}
    block = match.group(1)
    fields: dict[str, str] = {}
    for key, raw in FIELD_RE.findall(block):
        value = raw.strip()
        if (value.startswith('"') and value.endswith('"')) or (
            value.startswith("'") and value.endswith("'")
        ):
            value = value[1:-1]
        fields[key] = value
    return fields


def truncate(text: str, limit: int = 200) -> str:
    text = " ".join(text.split())
    if len(text) <= limit:
        return text
    return text[: limit - 1].rstrip() + "…"


def fallback_description(text: str) -> str:
    body = FRONTMATTER_RE.sub("", text, count=1).strip()
    for line in body.splitlines():
        line = line.strip()
        if line.startswith("#"):
            return truncate(line.lstrip("#").strip())
        if line:
            return truncate(line)
    return "(no description)"


def normalize_rel(path: Path, root: Path) -> str:
    return path.relative_to(root).as_posix()


def resolve_subagent_id(path: Path) -> str:
    stem = path.stem
    if stem.startswith("specialist-"):
        return stem[len("specialist-") :]
    return stem


def find_skill_link(text: str) -> str:
    match = SKILL_LINK_RE.search(text)
    if not match:
        return "-"
    skill_path = match.group(0)
    # extract folder name before SKILL.md as hint; full parse would need file read
    parts = skill_path.replace("\\", "/").split("/")
    if "SKILL.md" in parts:
        idx = parts.index("SKILL.md")
        if idx >= 1:
            return parts[idx - 1]
    return "-"


def collect_skills(cursor_dir: Path, root: Path) -> list[dict[str, str]]:
    rows: dict[str, dict[str, str]] = {}
    for skill_file in sorted(cursor_dir.glob("skills/**/SKILL.md")):
        text = skill_file.read_text(encoding="utf-8", errors="replace")
        meta = parse_frontmatter(text)
        skill_id = meta.get("name") or skill_file.parent.name
        rel = normalize_rel(skill_file, root)
        desc = meta.get("description") or fallback_description(text)
        rows[skill_id] = {
            "id": f"/{skill_id}",
            "path": normalize_rel(skill_file, root),
            "description": truncate(desc),
        }
    return sorted(rows.values(), key=lambda r: r["id"].lower())


def collect_commands(cursor_dir: Path, root: Path) -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    commands_dir = cursor_dir / "commands"
    if not commands_dir.is_dir():
        return rows
    for cmd_file in sorted(commands_dir.glob("*.md")):
        text = cmd_file.read_text(encoding="utf-8", errors="replace")
        meta = parse_frontmatter(text)
        cmd_id = meta.get("name") or cmd_file.stem
        rel = normalize_rel(cmd_file, root)
        desc = meta.get("description") or fallback_description(text)
        rows.append(
            {
                "id": f"/{cmd_id}",
                "path": normalize_rel(cmd_file, root),
                "description": truncate(desc),
                "linked": find_skill_link(text),
            }
        )
    return sorted(rows, key=lambda r: r["id"].lower())


def collect_subagents(cursor_dir: Path, root: Path) -> list[dict[str, str]]:
    seen_paths: set[str] = set()
    rows: dict[str, dict[str, str]] = {}
    agents_dir = cursor_dir / "agents"
    patterns = ("*.md", "*.mdc")
    files: list[Path] = []
    for pattern in patterns:
        files.extend(agents_dir.glob(pattern))
    for agent_file in sorted(files, key=lambda p: p.as_posix().lower()):
        rel = normalize_rel(agent_file, root)
        rel_key = rel.lower()
        if rel_key in seen_paths:
            continue
        seen_paths.add(rel_key)

        text = agent_file.read_text(encoding="utf-8", errors="replace")
        meta = parse_frontmatter(text)
        agent_id = meta.get("name") or resolve_subagent_id(agent_file)
        desc = meta.get("description") or fallback_description(text)
        rows[agent_id] = {
            "id": agent_id,
            "path": normalize_rel(agent_file, root),
            "description": truncate(desc),
            "invoke": "Task `subagent_type`",
        }
    return sorted(rows.values(), key=lambda r: r["id"].lower())


def render_table(headers: list[str], rows: list[list[str]]) -> str:
    if not rows:
        return "_None found._\n"
    lines = [
        "| " + " | ".join(headers) + " |",
        "| " + " | ".join("---" for _ in headers) + " |",
    ]
    for row in rows:
        escaped = [c.replace("|", "\\|") for c in row]
        lines.append("| " + " | ".join(escaped) + " |")
    return "\n".join(lines) + "\n"


def build_markdown(root: Path) -> str:
    cursor_dir = root / ".cursor"
    skills = collect_skills(cursor_dir, root)
    commands = collect_commands(cursor_dir, root)
    subagents = collect_subagents(cursor_dir, root)

    parts = [
        "# Cursor skills and subagents inventory",
        "",
        f"> Generated {date.today().isoformat()} from `.cursor/`",
        "",
        "## Skills",
        "",
        render_table(
            ["ID", "Path", "Description"],
            [[r["id"], r["path"], r["description"]] for r in skills],
        ),
        "## Slash commands",
        "",
        render_table(
            ["ID", "Path", "Description", "Linked skill"],
            [
                [r["id"], r["path"], r["description"], r["linked"]]
                for r in commands
            ],
        ),
        "## Subagents",
        "",
        render_table(
            ["ID", "Path", "Description", "Invoke via"],
            [
                [r["id"], r["path"], r["description"], r["invoke"]]
                for r in subagents
            ],
        ),
        "## Summary",
        "",
        f"- **Skills:** {len(skills)}",
        f"- **Slash commands:** {len(commands)}",
        f"- **Subagents:** {len(subagents)}",
        f"- **Scope:** `.cursor/skills/`, `.cursor/commands/`, `.cursor/agents/`",
        "",
    ]
    return "\n".join(parts)


def main() -> int:
    parser = argparse.ArgumentParser(description="Emit .cursor inventory tables")
    parser.add_argument(
        "--root",
        type=Path,
        default=None,
        help="Repository root (default: inferred from script location)",
    )
    parser.add_argument(
        "--out",
        type=Path,
        default=None,
        help="Write markdown to this file instead of stdout",
    )
    args = parser.parse_args()
    root = (args.root or repo_root_from_script()).resolve()
    if not (root / ".cursor").is_dir():
        print(f"error: no .cursor at {root}", file=sys.stderr)
        return 1
    md = build_markdown(root)
    if args.out:
        args.out.parent.mkdir(parents=True, exist_ok=True)
        args.out.write_text(md, encoding="utf-8")
        print(f"Wrote {args.out}", file=sys.stderr)
    else:
        print(md, end="")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
